incubator/hnc/internal/reconcilers/hierarchy_config.go

```diff
@@ -93,32 +95,33 @@ func (r *HierarchyConfigReconciler) Reconcile(req ctrl.Request) (ctrl.Result, er
 }
 
 func (r *HierarchyConfigReconciler) reconcile(ctx context.Context, log logr.Logger, nm string) error {
+	// Load the namespace and make a copy
 	nsInst, err := r.getNamespace(ctx, nm)
 	if err != nil {
-		if errors.IsNotFound(err) {
+		if apierrors.IsNotFound(err) {
 			// The namespace doesn't exist or is purged. Update the forest and exit.
 			// (There must be no HC instance and we cannot create one.)
 			r.onMissingNamespace(log, nm)
 			return nil
 		}
 		return err
 	}
+	origNS := nsInst.DeepCopy()
+
 	// Get singleton from apiserver. If it doesn't exist, initialize one.
 	inst, err := r.getSingleton(ctx, nm)
 	if err != nil {
 		return err
 	}
```

Comment:
Do we need to have the CRD being deleted case to return nil to not trigger a retry here?

Reply:
What did you have in mind if the CRD's been fully deleted?

Reply:
It seemed to me that when the CRD's been fully deleted, it will keep retrying even if the object is gone and reconciler watch won't be triggered. It will keep retrying until the reconciler is torn down, right?

Do you mean it's okay to have it keep retrying until the reconciler is down?

Reply:
I think so? I mean, what's the alternative? We could `os.Exit(0)` but then the pod will just start getting restarted, which isn't really much better.

Reply:
Oh I mentioned returning `nil` in the first comment.

Reply:
What about returning `nil, nil` in `getSingleton()` if the CRD is deleted and here if getting `nil, nil`, return `nil` instead of error to avoid retry? However, if doing so, and if the getting CRD just got a transient error, it won't retry while it should. Hmm, it seems too compllicated.

My only question to keep retrying is if the CRD is removed while HNC is actually not going to tear down, then the retry will never ends? BTW do we need to add webhook for users not to delete CRDs when HNC is supposed to be running?

Reply:
Ah, I see what you're saying. Yes, technically the reconciler could just early-exit if the CRD is not-found or is being deleted, but I don't mind if it spins for a while either and emits a lot of error messages (you _should_ see a lot of error messages if the pod isn't destroyed right away).

The retries have an exponential backoff so they'll get less and less frequent. I think that's good behaviour.

No, we shouldn't add a webhook to prevent this IMO. Deleting CRDs is a _very_ highly privileged operation, and it makes HNC _much_ harder to uninstall. I think we're fine for now, we can always come back to this later.

Reply:
Thanks for the explanation, sgtm!